docs/backend-system/building-plugins-and-modules/02-testing.md

```diff
@@ -6,21 +6,162 @@ sidebar_label: Testing
+
+The returned server also has a `port()` method which returns the dynamically
+bound listening port. You can use this to perform lower level network
+interactions with the running test service.
+
+## Testing Remote Service Interactions
+
+If your backend plugin or service interacts with external services using HTTP
+calls, we recommend leveraging the `msw` package to intercept actual outgoing
+requests and return mock responses. This lets you stub out remote services
+rather than the local clients, leading to more thorough and robust tests. You
+can read more about how it works [in their documentation](https://mswjs.io/).
+
+The `@backstage/backend-test-utils` package exports a `setupRequestMockHandlers`
+function which ensures that the correct `jest` lifecycle hooks are invoked to
+set up and tear down your `msw` instance, and enables the option that completely
+rejects requests that don't match one of your mock rules. This ensures that your
+tests cannot accidentally leak traffic into production from tests.
+
+Example:
+
+```ts
+import { setupRequestMockHandlers } from '@backstage/backend-test-utils';
+import { rest } from 'msw';
+import { setupServer } from 'msw/node';
+
+describe('read from remote', () => {
+  const worker = setupServer();
+  setupRequestMockHandlers(worker);
+
+  it('should auth and read successfully', async () => {
+    expect.assertions(1);
+
+    worker.use(
+      rest.get('https://remote-server.com/api/v3/foo', (req, res, ctx) => {
+        expect(req.headers.get('authorization')).toBe('Bearer fake');
+        return res(
+          ctx.status(200),
+          ctx.set('Content-Type', 'application/json'),
+          ctx.body(JSON.stringify({ value: 7 })),
+        );
+      }),
+    );
+
+    // exercise your plugin or service as usual, with real clients
   });
 });
 ```
+
+## Testing Database Interactions
+
+The `@backstage/backend-test-utils` package includes facilities for testing your
+plugins' interactions with databases, including spinning up `testcontainers`
+powered Docker images with real database engines to connect to.
+
+The base setup for such a test could look as follows:
+
+```ts
+// MyDatabaseClass.test.ts
+import { TestDatabaseId, TestDatabases } from '@backstage/backend-test-utils';
+import {
+  MyDatabaseClass,
+  applyDatabaseMigrations,
+  type FooTableRow,
+} from './MyDatabaseClass';
+
+describe('MyDatabaseClass', () => {
+  // Change this to the set of constants that you actually actively intend to
+  // support. Make sure to create only one TestDatabases instance per file,
+  // since spinning up "physical" databases to test against is much costlier
+  // than creating the "logical" databases within them that the individual
+  // tests use.
+  const databases = TestDatabases.create({
+    ids: ['POSTGRES_13', 'POSTGRES_9', 'SQLITE_3'],
+  });
+
+  // Just an example of how to conveniently bundle up the setup code
+  async function createSut(databaseId: TestDatabaseId) {
```

Comment:
I think "subject" might be a more commonly understood term?

Reply:
I dunno honestly, SUT is the one I've heard 🤷 :hahabusiness: :veryjava:

Reply:
Changed it to subject though, that reads better and feels less corporate :)